earn/src/data/hooks/UseRedeem.ts

```diff
@@ -0,0 +1,167 @@
+import { PermitState, usePermit } from 'shared/lib/data/hooks/UsePermit';
+import {
+  Address,
+  erc4626ABI,
+  useContractRead,
+  useContractReads,
+  useContractWrite,
+  usePrepareContractWrite,
+} from 'wagmi';
+
+import { ALOE_II_ROUTER_ADDRESS } from '../constants/Addresses';
+
+export enum RedeemState {
+  WAITING_FOR_INPUT,
+  FETCHING_DATA,
+  READY_TO_SIGN,
+  ASKING_USER_TO_SIGN,
+  READY_TO_REDEEM,
+  ASKING_USER_TO_REDEEM,
+}
+
+const PERMIT_STATE_TO_REDEEM_STATE = {
+  [PermitState.FETCHING_DATA]: RedeemState.FETCHING_DATA,
+  [PermitState.READY_TO_SIGN]: RedeemState.READY_TO_SIGN,
+  [PermitState.ASKING_USER_TO_SIGN]: RedeemState.ASKING_USER_TO_SIGN,
+  [PermitState.ERROR]: undefined,
+  [PermitState.DONE]: undefined,
+  [PermitState.DISABLED]: undefined,
+};
+
+const BN0 = BigNumber.from('0');
+
+export function useRedeem(chainId: number, lender: Address, amount: GN, owner: Address, recipient?: Address) {
+  if (!recipient) recipient = owner;
+
+  const erc4626 = {
+    address: lender,
+    abi: erc4626ABI,
+    chainId,
+  };
+
+  const router = {
+    address: ALOE_II_ROUTER_ADDRESS,
+    abi: routerABI,
+    chainId,
+  };
+
+  /*//////////////////////////////////////////////////////////////
+                              FETCHING
+  //////////////////////////////////////////////////////////////*/
+
+  const { data: maxData, isFetching: isFetchingMaxData } = useContractReads({
+    contracts: [
+      { ...erc4626, functionName: 'maxWithdraw', args: [owner] },
+      { ...erc4626, functionName: 'maxRedeem', args: [owner] },
+      { ...router, functionName: 'isMaxRedeemDynamic', args: [lender, owner] },
+    ] as const,
+    allowFailure: false,
+  });
+  const [maxAmount, maxShares, maxSharesIsChanging] = maxData ?? [BN0, BN0, false];
+
+  const { data: sharesData, isFetching: isFetchingShares } = useContractRead({
+    ...erc4626,
+    functionName: 'convertToShares',
+    args: [amount.toBigNumber()],
+    enabled: amount.toBigNumber().lte(maxAmount),
+  });
+  const shares = sharesData ?? BN0;
+
+  const threshold = maxShares.mul(95).div(100);
+  const shouldUseChecks = shares.gte(threshold) && maxSharesIsChanging;
+
+  /*//////////////////////////////////////////////////////////////
+                            ERC4626 REDEEM
+  //////////////////////////////////////////////////////////////*/
+
+  const redeemableShares = shares.lt(maxShares) ? shares : maxShares;
+  const { config: configRedeem } = usePrepareContractWrite({
+    ...erc4626,
+    functionName: 'redeem',
+    args: [redeemableShares, owner, recipient],
+    enabled: !shouldUseChecks && redeemableShares.gt(0),
+  });
+  const {
+    write: redeem,
+    data: redeemTxn,
+    isLoading: isAskingUserToRedeem,
+    reset: resetRedeemTxn,
+  } = useContractWrite({
+    ...configRedeem,
+    request: configRedeem.request
+      ? {
+          ...configRedeem.request,
+          gasLimit: configRedeem.request?.gasLimit.mul(110).div(100),
```

Comment:
Could we make the 110 into a constant with the same name we usually use? I plan to replace those with a universal constant, and making it a constant would make it easier for me to find it.